src/plugins/janus_sip.c

```diff
@@ -1801,6 +1801,95 @@ static void janus_sip_sofia_logger(void *stream, char const *fmt, va_list ap) {
 	}
 }
 
+static void janus_sip_sofia_logger_siptrace_single_log_callback(void *stream, char const *fmt, va_list ap) {
+	/* Since the fmt format string in the current Sofia SIP tport_log_msg function implementation is just "%s\n",
+	 * it's more efficient to directly work with the siptrace buffer. */
+	char *buffer = va_arg(ap, char *);
+
+	/* Check if this is a message we need */
+	if(buffer[0] == 's' && buffer[1] == 'e' && buffer[2] == 'n' && buffer[3] == 'd' && buffer[4] == ' ') {
+		/* An outgoing message is going to be logged, prepare for that */
+		int length = atoi(&buffer[5]);
+		JANUS_LOG(LOG_HUGE, "Intercepting message (%d bytes)\n", length);
+		/* Skip the stamp line */
+		buffer = strchr(buffer, '\n');
+		if(!buffer) {
+			JANUS_LOG(LOG_WARN, "No newline after the stamp in this message, dropping it...\n");
+			return;
+		}
+		buffer++;
+		/* Skip the separator line */
+		buffer = strchr(buffer, '\n');
+		if(!buffer) {
+			JANUS_LOG(LOG_WARN, "No newline after the separator in this message, dropping it...\n");
+			return;
+		}
+		buffer++;
+		/* If this is an OPTIONS, we don't care: drop it */
+		if(strstr(buffer, "OPTIONS") == buffer)
+			return;
+		/* Search for the Call-ID header and extract its value if it exists */
+		{
+			char *from = strstr(buffer, "Call-ID: ");
+			if(from) {
+				char *start = from + 9;
+				if(start) {
+					char *end = strchr(start, '\n');
+					if(end) {
+						size_t count = end - start;
+						count = count >= sizeof(call_id) ? sizeof(call_id) - 1 : count;
+						memcpy(call_id, start, count);
+						call_id[count] = '\0';
+					}
+				}
+			}
+		}
+		/* Look for the session this message belongs to */
+		janus_sip_session *session = NULL;
+		janus_mutex_lock(&sessions_mutex);
+		if(strlen(call_id))
+			session = g_hash_table_lookup(callids, call_id);
+		if(!session) {
+			/* Couldn't find any SIP session with that Call-ID, check the request */
+			if(strstr(buffer, "REGISTER") == buffer || strstr(buffer, "SIP/2.0 ") == buffer) {
+				/* FIXME This is a REGISTER or a response code:
+				* check the To header and get the identity from there */
+				char *from = strstr(buffer, "To: ");
+				if(from) {
+					char *start = strstr(from + 4, "<");
+					if(start) {
+						start++;
+						char *end = strstr(from, ">");
+						if(end) {
+							size_t count = end - start;
+							count = count >= sizeof(call_id) ? sizeof(call_id) - 1 : count;
+							memcpy(call_id, start, count);
+							call_id[count] = '\0';
+
```

Comment:
Unneeded empty line here.

Reply:
Sure, I'll remove this empty line.
I had originally used it to visually separate these two code blocks.